### PR TITLE
Correct GetCpmStringValue's second return value

### PR DESCRIPTION
**Pull request by guscarreon**
Function `GetCpmStringValue` found in `exchange/price_granularity.go` has been written to return `string` and `error` but currently does not really return a non-`nil` error under any scenario.  This PR adds an error value for `GetCpmStringValue` to return when a large `config.Precision` is found to avoid triggering a Panic. It also limits said value to a maximum of 4. Do you agree? Should we have more than 4 decimal figures in the returned CPM string?

Do you believe we should also return error under any other corner case listed in `TestGetPriceBucketString`?

### CODE REVIEW

File: exchange/price_granularity.go [hunk 1]
```diff
@@ -15,21 +16,27 @@ func GetCpmStringValue(cpm float64, config openrtb_ext.PriceGranularity) (string
 	bucketMax := 0.0
 	increment := 0.0
 	precision := config.Precision
-	// calculate max of highest bucket
+
+	// Limit the number of decimal significant figures. Very large values lead to Panics
+	if precision > 4 {
```
Comment (SyntaxNode) [inline exchange/price_granularity.go:21]: This needs review from @camrice and/or @VeronikaSolovei9. Can you please explain how large values lead to panics?

File: exchange/price_granularity.go [hunk 2]
```diff
@@ -15,21 +16,27 @@ func GetCpmStringValue(cpm float64, config openrtb_ext.PriceGranularity) (string
 	bucketMax := 0.0
```
Comment (SyntaxNode) [inline exchange/price_granularity.go:16]: While we're here, consider renaming the method to `GetPriceBucket` for improved clarity.

File: exchange/price_granularity.go [hunk 3]
```diff
@@ -15,21 +16,27 @@ func GetCpmStringValue(cpm float64, config openrtb_ext.PriceGranularity) (string
...
-	// calculate max of highest bucket
+
+	// Limit the number of decimal significant figures. Very large values lead to Panics
+	if precision > 4 {
+		return cpmStr, fmt.Errorf("Limit the number of precision figures to 4. Parsed value: %d", precision)
+	}
+
 	for i := 0; i < len(config.Ranges); i++ {
-		if config.Ranges[i].Max > bucketMax {
+		// calculate max of highest bucket
```
Comment (SyntaxNode) [inline exchange/price_granularity.go:26]: Consider removing this comment. I think the logic combined with the name `bucketMax` is self explanatory.

File: exchange/price_granularity_test.go [hunk 4]
```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
...
+		expectedPriceBucket string
+		expectedError       error
+	}
+	testGroups := []struct {
+		groupDesc string
+		inCpm     float64
+		testCases []aTest
+	}{
+		{
+			"cpm below the max in every price bucket",
```
Comment (SyntaxNode) [inline exchange/price_granularity_test.go:47]: The values here are not self descriptive enough. Please add field labels.

File: exchange/price_granularity_test.go [hunk 5]
```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
...
-
-func getOnePriceBucket(t *testing.T, name string, granularity openrtb_ext.PriceGranularity, price float64, expected string) {
-	t.Helper()
-	priceBucket, err := GetCpmStringValue(price, granularity)
-	if err != nil {
-		t.Errorf("Granularity: %s :: GetPriceBucketString: %s", name, err.Error())
+	// Define test cases
+	type aTest struct {
+		granularityId       string
+		inGranularity       openrtb_ext.PriceGranularity
```
Comment (SyntaxNode) [inline exchange/price_granularity_test.go:37]: Rename to `granularity`?

File: exchange/price_granularity_test.go [hunk 6]
```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
...
+		expectedPriceBucket string
+		expectedError       error
+	}
+	testGroups := []struct {
+		groupDesc string
+		inCpm     float64
```
Comment (SyntaxNode) [inline exchange/price_granularity_test.go:43]: Rename to `cpm`?

File: exchange/price_granularity_test.go [hunk 7]
```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
...
+				{"dense", dense, "5.70", nil},
+				{"custom1", custom1, "5.70", nil},
+			},
+		},
+		{
+			"Precision value corner cases",
+			1.876,
+			[]aTest{
+				{
+					"Negative precision defaults to number of digits already in CPM float",
```
Comment (SyntaxNode) [inline exchange/price_granularity_test.go:75]: I find the tests hard to reach with a long granularity id. I don't understand the setup.

Comment (hhhjort) [inline exchange/price_granularity_test.go:75]: Where is the code that deals with a negative precision?

File: exchange/price_granularity_test.go [hunk 8]
```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
...
+				},
+				{
+					"Precision value equals zero, we expect to round up to the nearest integer",
+					openrtb_ext.PriceGranularity{Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
+					"2",
+					nil,
+				},
+				{
+					"Very large precision value: expect error and an empty string in return",
+					openrtb_ext.PriceGranularity{Precision: int(^uint(0) >> 1), Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
```
Comment (hhhjort) [inline exchange/price_granularity_test.go:88]: Do we really want to test with about the largest possible integer when the limit is only 4? I would worry that this could potentially generate an error due to being silly large rather than violating our stated limit of 4. Would probably  generate a different error message, or error out in another way, so would still be flagged, but doesn't make sense as the number doesn't really compare to 4.

File: exchange/price_granularity.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply exchange/price_granularity.go:21]: I tested current upstream implementation with precision = 10 and it works. 
The only thing it looks scary in response: 
```
{
                    "hb_pb": "17.0000000000",
                    "hb_pb_cat_dur": "17.0000000000_food_30s",
                    "hb_cache_id": "37be5f83-4be9-41b4-ab3f-84883205591d"
                }
```

Comment (guscarreon) [reply exchange/price_granularity.go:21]: @VeronikaSolovei9 do you think we should should put a limit to `precision`? If so, what should it be?

Comment (VeronikaSolovei9) [reply exchange/price_granularity.go:21]: Maybe we should not return error in this case, instead just overwrite precision with maximum reasonable value?

Comment (VeronikaSolovei9) [reply exchange/price_granularity.go:21]: @guscarreon 2 and 4 sounds reasonable to me, but please confirm it with product owners as well.

Comment (SyntaxNode) [reply exchange/price_granularity.go:21]: In this case, the product owners are the Prebid Server Committee. This is already configurable by the publisher. What's the worst thing that would happen if they enter a value too large? Is this really something we need to code logic for?

Comment (VeronikaSolovei9) [reply exchange/price_granularity.go:21]: "hb_pb_cat_dur" is a targeting key for Ad server to select Line Items. Redirect URL also contains it. With wrong precision Ad server set up will not work.

**Comment by SyntaxNode** [comment]
After further discussion, we learned the precision limit is intended to prevent the server from panicking due to a very high publisher supplied value. The check is necessary and will be moved out of this method (so it doesn't need to return an error) and included as part of the other request validation checks. This will not be configurable since the limit will be set to a high value.

File: exchange/price_granularity_test.go (hunk 7 above)
Comment (guscarreon) [reply exchange/price_granularity_test.go:75]: Function `strconv.FormatFloat(roundedCPM, 'f', precision, 64)` deals with negative precision values. More details in the function definition itself.

File: exchange/price_granularity_test.go (hunk 8 above)
Comment (guscarreon) [reply exchange/price_granularity_test.go:88]: You are totally right, that was too much. The latest version limits the precision value to 2^15 that both not panics and seems to not overflow the return string. I believe 2^15 = 32768 is more than enough. but let me know your take on this.

File: exchange/price_granularity.go [hunk 9]
```diff
@@ -7,33 +7,32 @@ import (
...
+func GetPriceBucket(cpm float64, config openrtb_ext.PriceGranularity) string {
 	cpmStr := ""
 	bucketMax := 0.0
 	increment := 0.0
 	precision := config.Precision
-	// calculate max of highest bucket
+
 	for i := 0; i < len(config.Ranges); i++ {
-		if config.Ranges[i].Max > bucketMax {
+		if bucketMax < config.Ranges[i].Max {
```
Comment (SyntaxNode) [inline exchange/price_granularity.go:18]: IMHO: I preferred it the other way. It think it reads better as "if this range max is higher than the bucketMax, set bucketMax to it".

File: exchange/price_granularity_test.go (hunk 7 above)
Comment (hhhjort) [reply exchange/price_granularity_test.go:74]: Are you testing that `strconv.FormatFloat()` is working properly, or that something is handling the case properly when `FormatFloat()` does what it is supposed to do, or that your code is subsequently handling that output correctly? I don't see any code that would react to anything particular about a negative number.

Comment (SyntaxNode) [reply exchange/price_granularity_test.go:74]: https://github.com/prebid/prebid-server/blob/550770752797f12a6336be551e991793ab89c591/openrtb_ext/request.go#L178

File: exchange/price_granularity.go (hunk 9 above)
Comment (guscarreon) [reply exchange/price_granularity.go:18]: Reversed

File: exchange/price_granularity_test.go (hunk 7 above)
Comment (guscarreon) [reply exchange/price_granularity_test.go:74]: Given that `PriceGranularity` has its own `func (pg *PriceGranularity) UnmarshalJSON(b []byte) error` inside `openrtb_ext/request.go` that sets a bunch of default values if found missing in the Json string, I thought it'd be better to cap the precision value there instead.

**Comment by VeronikaSolovei9** [comment]
I have nothing to add, lgtm

File: openrtb_ext/request.go [hunk 10]
```diff
@@ -192,9 +193,6 @@ func (pg *PriceGranularity) UnmarshalJSON(b []byte) error {
 			}
 			// Enforce that we don't read "min" from the request
 			pgraw.Ranges[i].Min = prevMax
-			if pgraw.Ranges[i].Min < prevMax {
-				return errors.New("Price granularity error: overlapping granularity ranges")
-			}
```
Comment (guscarreon) [inline openrtb_ext/request.go:194]: If we are ignoring `"min"` json field

Comment (SyntaxNode) [reply openrtb_ext/request.go:194]: Yeah. That condition will never be met with the `pgraw.Ranges[i].Min = prevMax` line right before it.

**Review by hhhjort — DISMISSED** [review]

File: exchange/price_granularity_test.go [hunk 11]
```diff
@@ -28,32 +31,105 @@ func TestGetPriceBucketString(t *testing.T) {
...
+					openrtb_ext.PriceGranularity{Precision: -1, Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
+					"1.85",
+				},
+				{
+					"Precision value equals zero, we expect to round up to the nearest integer",
+					openrtb_ext.PriceGranularity{Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
+					"2",
+				},
+				{
+					"Largest precision value PBS supports (2^15 = 32768)",
```
Comment (SyntaxNode) [inline exchange/price_granularity_test.go:84]: Do we want to change this to 15 from 32768 to match the other change?

Comment (guscarreon) [reply exchange/price_granularity_test.go:84]: Corrected

**Review by hhhjort — APPROVED** [review]

**Review by SyntaxNode — APPROVED** [review]
